virtual_eeprom/virtual_eeprom.c

```diff
@@ -0,0 +1,219 @@
+#include <linux/module.h>
+#include <linux/kernel.h>
+#include <linux/nvmem-provider.h>
+#include <linux/slab.h>
+#include <linux/mutex.h>
+#include <linux/of.h>
+#include <linux/list.h>
+#include <linux/fs.h>
+#include <linux/fcntl.h>
+#include <linux/err.h>
+#include <linux/uaccess.h>
+#include <linux/platform_device.h>
+
+// Structure to hold EEPROM data and metadata
+struct virtual_eeprom {
+    char *data;
+    struct nvmem_device *nvmem_dev;
+    struct nvmem_config nvmem_cfg;
+    struct mutex mutex;
+    int size;
+    const char *name;
+    const char *bin_path;
+    struct list_head list;
+    struct platform_device *pdev;
+};
+
+// Function prototypes
+static int save_eeprom_to_file(struct virtual_eeprom *eeprom);
+static int load_eeprom_from_file(struct virtual_eeprom *eeprom);
+
+// Initialize global list to track all EEPROMs
+static LIST_HEAD(eeprom_list);
+
+static int eeprom_read(void *context, unsigned int offset, void *val, size_t bytes)
+{
+    struct virtual_eeprom *eeprom = context;
+
+    if (offset >= eeprom->size)
+        return -EINVAL;
+
+    if (offset + bytes > eeprom->size)
+        bytes = eeprom->size - offset;
+
+    mutex_lock(&eeprom->mutex);
+    memcpy(val, eeprom->data + offset, bytes);
+    mutex_unlock(&eeprom->mutex);
+
+    return 0;
+}
+
+static int eeprom_write(void *context, unsigned int offset, void *val, size_t bytes)
+{
+    struct virtual_eeprom *eeprom = context;
+    int ret;
+
+    if (offset >= eeprom->size)
+        return -EINVAL;
+
+    if (offset + bytes > eeprom->size)
+        bytes = eeprom->size - offset;
+
+    mutex_lock(&eeprom->mutex);
+    memcpy(eeprom->data + offset, val, bytes);
+    // Save the data to the file after each write
+    ret = save_eeprom_to_file(eeprom);
+    mutex_unlock(&eeprom->mutex);
+
+    return ret < 0 ? ret : 0;
+}
+
+static int save_eeprom_to_file(struct virtual_eeprom *eeprom)
+{
+    struct file *file;
+    loff_t pos = 0;
+    ssize_t ret;
+
+    file = filp_open(eeprom->bin_path, O_WRONLY | O_CREAT | O_TRUNC, 0644);
+    if (IS_ERR(file)) {
+        pr_err("Failed to open file %s for writing: %ld\n", eeprom->bin_path, PTR_ERR(file));
+        return PTR_ERR(file);
+    }
+
+    ret = kernel_write(file, eeprom->data, eeprom->size, &pos);
```

Comment:
This should continue to try write the rest of the file if ret > 0 but < eeprom->size.
kernel_write may not write the all the data at once.

Reply:
same with read

Reply:
fixed